lib/light-service/organizer/iterate.rb

```diff
@@ -1,22 +1,82 @@
+# require 'light-service/action.rb'
+
 module LightService
   module Organizer
-    class Iterate
-      extend ScopedReducable
+    class Iterator
+      using ::LightService::Refinements::Array
+
+      include ScopedReducable
+
+      def initialize(collection_key, actions)
+        @collection_key = collection_key
+        @actions = actions
+        @iterations = nil
+        @current_iteration = nil
+      end
+
+      def rollback(ctx)
+        reversable_iterations.reverse.each_with_index do |single_iteration, idx|
+          ctx[single_iteration.item_key] = single_iteration.item
+
+          unless single_iteration.actions.include?(ctx.current_action)
+            ctx.current_action = single_iteration.actions.last
+          end
+
+          ctx.current_action = single_iteration.actions.last unless idx.zero?
+
+          SubOrganizer.new(self).with(ctx).reduce_rollback(single_iteration.actions)
+        end
+
+        ctx
+      end
+
+      def call(ctx)
+        return ctx if ctx.stop_processing?
+
+        setup_iterations(ctx, @collection_key, @actions)
+
+        @iterations.members.each do |single_iteration|
+          ctx[single_iteration.item_key] = single_iteration.item
+          @current_iteration = single_iteration
+          ctx = SubOrganizer.new(self).with(ctx).reduce(single_iteration.actions)
+        end
+
+        ctx
+      end
+
+      def self.run(_organizer, collection_key, actions)
+        new(collection_key, Array.wrap(actions))
+      end
+
+      private
+
+      def setup_iterations(ctx, collection_key, actions)
+        @iterations ||= Iterations.new(ctx, collection_key, actions)
+      end
 
-      def self.run(organizer, collection_key, steps)
-        lambda do |ctx|
-          return ctx if ctx.stop_processing?
+      def reversable_iterations
+        index_of_current_iteration = @iterations.members.index(@current_iteration) || 0
 
+        # Reverse from the point where the fail was triggered
+        @iterations.members.take(index_of_current_iteration + 1)
+      end
+
+      class Iterations
```

Comment:
This is the exploded and static representation of the whole iterations. Its `members` is and array of `SingleIteration`. Each `SingleIteration` brings its `item`, `item_key` (probably useless) and the whole set of `Action`s to be run within it.